### ISSUE TITLE
Webhooks cryptographic signature

### ISSUE THREAD

**Issue by bobf** [reapit/foundations#5607]
**Is your feature request related to a problem? Please describe.**
Webhooks do not appear to contain a cryptographic signature which means I cannot verify the authenticity of the payload.

**Describe the solution you'd like**
Reapit could publish a public key and sign all webhooks using the private counterpart which would allow me to verify authenticity of the received message against the public key.

**Describe alternatives you've considered**
N/A.

**Additional context**
N/A

**Specification**
- Understand how request signing is implemented in other systems, bearing in mind listeners will want to verify the source of the message (that it's come from Reapit) and also that it has not been tampered with. It may be useful to refer to documentation for other enterprise products (Stripe etc) to see how they handle this
- Outcome of this ticket is a list of options which we will discuss as a group before deciding on implementation. A POC could be included here so we can demo what this looks like (if appropriate)
- https://tools.ietf.org/id/draft-cavage-http-signatures-08.html

**Comment by HollyJoyPhillips** [comment]
Whilst the nature of the request has been accepted, we are unable to commit to a specified sprint and therefore will assign your issue to the ‘Long Term’ column. We will regularly review any issues and where development capacity is available, or work is aligned with our Roadmap, the issue will be updated.

**Comment by bobf** [comment]
@HollyJoyPhillips Hi, is there any movement on this or are you able to give us any kind of projection on when it might be implemented ?

This is something we see as being critical to our HFPE project.

Regards,

**Comment by plittlewood-rpt** [comment]
Hi @bobf - this is something that has been discussed a couple of times and is likely to be done, but not in the near term. You can read about how webhook endpoints can be secured in the [documentation here](https://foundations-documentation.reapit.cloud/api/webhooks#securing-your-endpoint) and whilst I appreciate it's not as secure as a public/private key implementation, it does add a layer of security and stops your endpoints being completely public.

**Comment by plittlewood-rpt** [comment]
Carrying to next sprint - no further work was completed on this last week due to staff absence. This is a fairly complex piece of work so may take some time to implement

**Comment by bobf** [comment]
@plittlewood-rpt Thanks for the update. We'll be starting our user data sync project soon, but we won't be able to go live without this feature - it will take a while before we're ready to launch so hopefully that will give enough time to get this ready.

**Comment by roscoe2001** [comment]
Just to add, we're following this with interest as it's something we would like as well before going live with some web hooks that we've done.

**Comment by plittlewood-rpt** [comment]
Thanks everyone for their input and interest. To give you a basic idea of how this is being implemented (subject to change)

- A public/private key will be created for each application (so multiple webhook configurations for the same app will share the same set of keys)
  - When a webhook is registered for an app for the first time, the key pair will be generated at this point.
  - When a subsequent webhook is registered for an app, the existing key pair for that app will be used
- The public key will be available at an endpoint where you will be able to access the public keys for your application(s)
- Any outbound webhook events will be signed using the private key , allowing you to verify the sender of the message. We will be using the Ed25519 algorithm instead of the traditional RSA approach as this is considered more secure than RSA these days

**Comment by bobf** [comment]
This sounds like a sensible implementation to me.

The key part for me is:

> The public key will be available at an endpoint where you will be able to access the public keys for your application(s)

This will allow us to automate the verification process in our general-purpose Foundations library without having to manually download files etc.

Thanks for being thorough on this.

**Comment by plittlewood-rpt** [comment]
Carried to current sprint. Work largely done but some automated tests still to write before E2E testing in dev environment

**Comment by roscoe2001** [comment]
Hi, is the an updated estimated delivery on this one please (I'm just putting a plan together which has a dependency on this piece of work). I don't know if it makes a difference but it's the Works Order web hooks that we need initially.

**Comment by plittlewood-rpt** [comment]
Hi @roscoe2001 - the backend work should go to production in the next couple of weeks, however there's some frontend work required also to allow you to access your public key via the developer portal.

**Comment by plittlewood-rpt** [comment]
This is now in dev and working well. To discuss frontend work with @HollyJoyPhillips

**Comment by plittlewood-rpt** [comment]
There is a separate ticket for production rollout which has some dependencies. Please follow #6855 for updates after this week.

**Comment by plittlewood-rpt** [comment]
To update ticket followers - this work has now been signed off in our development environment. There are some additional tasks that need to be completed before we can release to production. This ticket will now be closed and followers should follow #6855 for further updates

**Comment by plittlewood-rpt** [comment]
To update followers of this ticket, it is looking like this will all be enabled in production on Monday 6th June. Documentation surrounding signature verification will be available in the DeveloperPortal under Docs -> APIs -> Webhooks -> Securing your endpoint
